scripts/staging/fm_binclass.dml

```diff
@@ -0,0 +1,122 @@
+#-------------------------------------------------------------
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+# 
+#   http://www.apache.org/licenses/LICENSE-2.0
+# 
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+#-------------------------------------------------------------
+
+
+/*
+ * Factorization Machines for binary classification.
+ */
+
+# Imports
+source("nn/optim/adam.dml") as adam
+source("staging/fm.dml") as fm
+source("nn/layers/log_loss.dml") as log_loss
+source("nn/layers/sigmoid.dml") as sigmoid
+
+train = function(matrix[double] X)
```

Comment:
This function is looking good.  We do need to pass in the correct labels `y` as well to the function, as they are needed in order to compute the loss (i.e., how correct was the model in a probabilistic sense).  At this point, we have the setup steps (1 & 2), and the steps for a single update step (3-9).  Now, we just need to wrap steps 3-9 in a loop for some number of iterations (which could be passed in as a parameter) so that we can optimize the model over time.  Additionally, we should update the steps to use a mini-batch approach instead of a full-batch approach.  Basically (after steps 1 & 2), we'll need a loop that goes through the data in `X` and pulls out mini-batches (of some size that can be passed into this function), running steps 3-9 for each mini-batch.  Then, we can wrap that loop inside another loop that decides how many times to go through the full dataset, i.e., the number of `epochs`.

Roughly, it looks like this:
```
setup stuff (steps 1 & 2)

for some number of epochs:
  for each mini-batch `X_batch, y_batch` in `X, y`:
    steps 3-9 on the current mini-batch
```

The [`mnist_lenet.dml`](https://github.com/apache/systemml/blob/master/scripts/nn/examples/mnist_lenet.dml#L107) is a great example.